coderd/httpapi/httpapi.go

```diff
@@ -187,37 +185,50 @@ func ServerSentEventSender(rw http.ResponseWriter, r *http.Request) (func(ctx co
 		panic("http.ResponseWriter is not http.Flusher")
 	}
 
-	// Send a heartbeat every 15 seconds to avoid the connection being killed.
+	closed = make(chan struct{})
+	type sseEvent struct {
+		payload []byte
+		errC    chan error
+	}
+	eventC := make(chan sseEvent)
+
+	// Synchronized handling of events (no guarantee of order).
 	go func() {
+		defer close(closed)
+
+		// Send a heartbeat every 15 seconds to avoid the connection being killed.
 		ticker := time.NewTicker(time.Second * 15)
 		defer ticker.Stop()
 
 		for {
+			var event sseEvent
+
 			select {
 			case <-r.Context().Done():
 				return
```

Comment:
Why doesn't this prevent the write/flush after finish? We're not hijacking the connection so the context should still be cancelled on finish right?

Reply:
Because this code had the following race condition:

1. Say `watchWorkspace` encountered an error and exited (this begins the process of all handlers and middleware exiting (`r.Context()` is not yet cancelled)
2. Timer is triggered
3. We write to `rw` which still succeeds as the teardown is still happening
4. Teardown finishes, context is cancelled, but we're currently not listening to this signal
5. We hit `Flush` after teardown so we get a nil pointer deref

Reply:
Awesome explanation thank you